### PR TITLE
Add I18n.interpolation_keys

### PR DESCRIPTION
**Pull request by tom-lord**
Adds a **new method**, `I18n.interpolation_keys`, which returns a list of keys needed to fully resolve a translation.

Some examples:

```
    # Suppose we have the following:
    #   I18n.t 'example.zero' == 'Zero interpolations'
    #   I18n.t 'example.one' == 'One interpolation %{foo}'
    #   I18n.t 'example.two' == 'Two interpolations %{foo} %{bar}'
    #   I18n.t 'example.three' == ['One %{foo}', 'Two %{bar}', 'Three %{baz}']
    #   I18n.t 'example.one', locale: :other == 'One interpolation %{baz}'
    #
    # Then we can expect the following results:
    #   I18n.interpolation_keys('example.zero') #=> []
    #   I18n.interpolation_keys('example.one') #=> ['foo']
    #   I18n.interpolation_keys('example.two') #=> ['foo', 'bar']
    #   I18n.interpolation_keys('example.three') #=> ['foo', 'bar', 'baz']
    #   I18n.interpolation_keys('one', scope: 'example', locale: :other) #=> ['baz']
    #   I18n.interpolation_keys('does-not-exist') #=> []
    #   I18n.interpolation_keys('example') #=> []
```

---

In a recent project, I wanted to be able to figure out "What keys are needed for this translation"?
To my surprise, there wasn't a built-in method for this in `I18n`; the only option was to hand-roll some regex to look for `%{...}` patterns in the translation.

So now I've tried to implement it here "properly", making use of `I18n.config.interpolation_patterns` to be more robust with how interpolations are identified by the library -- e.g. it will also [match `%<foo>.d` by default](https://github.com/ruby-i18n/i18n/blob/74258ad03021c3592302438b86e08ab778e85007/lib/i18n/interpolate/ruby.rb#L10), and respect [custom pattern configurations like `{{foo}}`](https://github.com/ruby-i18n/i18n/blob/74258ad03021c3592302438b86e08ab778e85007/test/i18n/interpolate_test.rb#L103).

### CODE REVIEW

File: lib/i18n.rb [hunk 1]
```diff
@@ -231,6 +231,31 @@ def translate!(key, **options)
...
+    #   I18n.t 'example.one' == 'One interpolation %{foo}'
+    #   I18n.t 'example.two' == 'Two interpolations %{foo} %{bar}'
+    #   I18n.t 'example.one', locale: :other == 'One interpolation %{baz} %{bar}'
+    #
+    # Then we can expect the following results:
+    #   I18n.interpolation_keys('example.zero') #=> []
+    #   I18n.interpolation_keys('example.one') #=> ['foo']
+    #   I18n.interpolation_keys('example.two') #=> ['foo', 'bar']
+    #   I18n.interpolation_keys('one', scope: 'example', locale: :other) #=> ['baz']
+    #   I18n.interpolation_keys('does-not-exist') #=> []
```
Comment (tom-lord) [inline lib/i18n.rb:247]: ...Or should it be `nil` 🤷🏼‍♂️

Comment (radar) [reply lib/i18n.rb:247]: I'd pick an empty array here, to make the return type the same for all variants of this method.

**Comment by radar** [comment]
> In a recent project, I wanted to be able to figure out "What keys are needed for this translation"?

I'd be keen to know what this project is :) It sounds a little in the vein of LocaleApp, etc.

I think the feature is lightweight enough and it could be useful for others. Please do add some tests here.

File: lib/i18n.rb [hunk 2]
```diff
@@ -231,6 +231,34 @@ def translate!(key, **options)
...
+    def interpolation_keys(key, **options)
+      raise I18n::ArgumentError if !key.is_a?(String) || key.empty?
+
+      return [] unless exists?(key, **options.slice(:locale, :scope))
+
+      translation = translate(key, **options.slice(:locale, :scope))
+
+      raise I18n::NonStringTranslationError unless translation.is_a?(String)
```
Comment (tom-lord) [inline lib/i18n.rb:255]: This _could_ also return an empty `Array`? 🤔 

It's a bit of an odd one: What if you have translations:

```
example_nested:
  one: "%{foo}"
  two: %{bar}"
```

What should `I18n.interpolation_keys('example_nested')` return? `[]`, `[['foo'], ['bar']]`, raise an exception, or ... Something else? 👀

File: test/i18n_test.rb [hunk 3]
```diff
@@ -287,6 +287,34 @@ def setup
...
+    assert_equal ["foo", "bar"], I18n.interpolation_keys("example_two")
+  end
+
+  test "interpolation_keys returns an empty array when no interpolations " do
+    store_translations(:en, "example_zero" => "Zero interpolations")
+    assert_equal [], I18n.interpolation_keys("example_zero")
+  end
+
+  test "interpolation_keys returns an empty array when missing translation " do
+    assert_equal [], I18n.interpolation_keys("does-not-exist")
```
Comment (tom-lord) [inline test/i18n_test.rb:301]: This _could_ raise an exception or return `nil`, but in the interest of simplicity and consistency, I feel it's probably fine to return `[]`.

File: test/i18n_test.rb [hunk 4]
```diff
@@ -287,6 +287,34 @@ def setup
...
+  end
+
+  test "interpolation_keys returns an empty array when nested translation" do
+    store_translations(:en, "example_nested" => { "one" => "One %{foo}", "two" => "Two %{bar}" })
+    assert_equal [], I18n.interpolation_keys("example_nested")
```
Comment (tom-lord) [inline test/i18n_test.rb:306]: In this scenario, `I18n.t("example_nested")` _does not require/use any interpolation keys_ (even though the child translations do). Therefore I feel it's correct for the method to return `[]`.

File: lib/i18n.rb [hunk 5]
```diff
@@ -429,6 +453,17 @@ def normalize_key(key, separator)
           keys
         end
     end
+
+    def interpolation_keys_from_translation(translation)
+      case translation
+      when ::String
+        translation.scan(Regexp.union(I18n.config.interpolation_patterns))
+      when ::Array
+        translation.map { |element| interpolation_keys_from_translation(element) }
```
Comment (tom-lord) [inline lib/i18n.rb:462]: Note that this [does need to be recursive](https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/backend/base.rb#L183-L198)!

**Comment by tom-lord** [comment]
@radar Sorry for the slow progress, I was on holiday 😄 

I've gone through some more edge cases, added more documentation and tests. Please take a look when you have time!

**Comment by tom-lord** [comment]
> I'd be keen to know what this project is :) It sounds a little in the vein of LocaleApp, etc.

I did have applications like that in mind when proposing to push this feature into the main `I18n` project. However, my specific problem was a bit more granular:

In an application, translations may or may not utilise interpolation keys, in each country. Sometimes, resolving those keys is an "expensive" operation. So as an optimisation, it makes sense to skip trying to resolve any keys that aren't actually being used.

File: test/i18n_test.rb (hunk 3 above)
Comment (radar) [reply test/i18n_test.rb:301]: I think `[]` is better here as it makes the return type consistently `Array` rather than `Array | nil`

**Comment by radar** [comment]
Thanks!
